docs/rfcs/0004-rfc-connectiomanager.md

```diff
@@ -0,0 +1,76 @@
+- Status: In progress
+- Date: 2023-11-02
+- Author: [@gela](https://github.com/gela)
+- RFC PR: TBD
+- Implementation GitHub issue: TBD.
+
+
+## Summary
+
+Propose a new `ConnectionManager` component which handles all connectivity related checks once during VPN activation (`Controller::StateConnecting`) and then periodically once the VPN is already activated (`Controller::StateOn`); the controller will continue to exist as is, communicating with the frontend as the backend goes through various states such as `connecting` and `confirming` to reflect those to the user.
+
+## Motivation
+
+At present, the handling of VPN connectivity lacks standardization and consistency. There are instances when the VPN goes into No Signal due to a variety of connectivity-related issues, such as the network connection failing or the server location becoming unavailable. In these situations, users are often left without a clear explanation of what went wrong or guidance on how to troubleshoot and resolve the issue. They may encounter a background error message or lose signal without further information.
+
+The objective with the Connection Manager is to address these issues by running probes on all potential causes and presenting the user with an appropriate modal detailing the root cause of their problem. This approach aims to enhance the user's understanding of the issue and provide them with actionable steps to rectify it.
+
+To avoid further complicating the already complex logic of the controller, this document proposes encapsulating this new functionality within the Connection Manager. This object will interact with both the controller and connection health, serving as a dedicated entity for managing and improving VPN connectivity. This approach ensures a more streamlined and efficient handling of VPN connectivity issues, ultimately enhancing the user experience.
+
+## Proposed Solution
+
+We will introduce a new object, the Connection Manager, which will be responsible for managing all probes. These probes are stored as distinct states within an enumeration class, `ConnectionManager::State`. The Connection Manager will sequentially cycle through each of these states in a systematic order, only proceeding to the next state if the current probe is successful.
```

Comment:
> only proceeding to the next state if the current probe is successful.

> Upon encountering a failed probe, the progression through the remaining probes is immediately halted.

We'll still need to do some diagnostic work though right? Unless we work out a perfect sequence by which to run checks, it seems possible that we could fail at `check A` when the failure is actually a failing `check B`, and, without further detective work, emit the wrong error message.

Reply:
What sort of diagnostics are you thinking? I was under the impression that we would identify the cause and then show the user some things they can try. For example if the network check fails, we'll tell them to check their internet. I'm not sure if there's much (or anything really) we as the VPN client can do to resolve any of the issues the connection manager identifies.

Reply:
Resolving this because it's outdated and I'm struggling to load 100% of the context for this comment back into memory.  I think what I was getting at is the possibility of sending the wrong error message to the user when, for random example, when running checks on activation, we fail at the subscription check but the reason for the failure is actually that there is no internet or a given server is unavailable.